skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -0,0 +1,430 @@
+"""
+This is an implementation of the 2D/3D thinning algorithm
+of [Lee94]_ of binary images, based on [IAC15]_. 
+
+The original Java code [IAC15]_ carries the following message:
+
+ * This work is an implementation by Ignacio Arganda-Carreras of the
+ * 3D thinning algorithm from Lee et al. "Building skeleton models via 3-D 
+ * medial surface/axis thinning algorithms. Computer Vision, Graphics, and 
+ * Image Processing, 56(6):462-478, 1994." Based on the ITK version from
+ * Hanno Homann <a href="http://hdl.handle.net/1926/1292"> http://hdl.handle.net/1926/1292</a>
+ * <p>
+ *  More information at Skeletonize3D homepage:
+ *  http://fiji.sc/Skeletonize3D
+ *
+ * @version 1.0 11/13/2015 (unique BSD licensed version for scikit-image)
+ * @author Ignacio Arganda-Carreras (iargandacarreras at gmail.com)
+
+References
+----------
+.. [Lee94] T.-C. Lee, R.L. Kashyap and C.-N. Chu, Building skeleton models
+       via 3-D medial surface/axis thinning algorithms.
+       Computer Vision, Graphics, and Image Processing, 56(6):462-478, 1994.
+
+.. [IAC15] Ignacio Arganda-Carreras, 2015. Skeletonize3D plugin for ImageJ(C).
+           http://fiji.sc/Skeletonize3D
+
+"""
+from __future__ import division, print_function, absolute_import
+
+from libc.string cimport memcpy
+
+import numpy as np
+from numpy cimport npy_intp, npy_uint8, ndarray
+cimport cython
+
+ctypedef npy_uint8 pixel_type
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _compute_thin_image(pixel_type[:, :, ::1] img not None):
+    """Compute a thin image.
+
+    Loop through the image multiple times, removing "simple" points, i.e.
+    those point which can be removed without changing local connectivity in the
+    3x3x3 neighborhood of a point.
+
+    This routine implements the two-pass algorithm of [Lee94]_. Namely,
+    for each of the six border types (positive and negative x-, y- and z-),
+    the algorithm first collects all possibly deletable points, and then
+    performs a sequential rechecking.
+
+    The input, `img`, is assumed to be a 3D binary image in the
+    (p, r, c) format [i.e., C ordered array], filled by zeros (background) and
+    ones. Furthermore, `img` is assumed to be padded by zeros from all
+    directions --- this way the zero boundary conditions are automatic
+    and there is need to guard against out-of-bounds access.
+
+    """
+    cdef:
+        int unchanged_borders = 0, curr_border, num_borders
+        int borders[6]
+        npy_intp p, r, c
+        bint no_change
+
+        list simple_border_points
+        Py_ssize_t num_border_points, i, j
+        (npy_intp, npy_intp, npy_intp) point
+
+        pixel_type neighb[27]
+    borders[:] = [4, 3, 2, 1, 5, 6]
```

Comment:
Pixel comparisons with ImageJ do not pass otherwise. Something to do with xyz vs prc, I suppose.